alpha/alpha-core/src/main/java/org/apache/servicecomb/saga/alpha/core/TxConsistentService.java

```diff
@@ -44,7 +58,32 @@ public void handle(TxEvent event) {
   }
 
   private void compensate(TxEvent event) {
-    List<TxEvent> events = eventRepository.findStartedTransactions(event.globalTxId(), EventType.TxStartedEvent.name());
+    List<TxEvent> events = eventRepository.findStartedTransactions(event.globalTxId(), TxStartedEvent.name());
     events.forEach(omegaCallback::compensate);
+    eventsToCompensate.computeIfAbsent(event.globalTxId(), (v) -> {
+      CopyOnWriteArraySet<String> eventSet = new CopyOnWriteArraySet<>();
+      events.forEach(e -> eventSet.add(getUniqueEventId(e)));
+      return eventSet;
+    });
+  }
+
+  private void updateCompensateStatus(TxEvent event) {
+    Set<String> events = eventsToCompensate.get(event.globalTxId());
+    if (events != null) {
+      events.remove(getUniqueEventId(event));
+      if (events.isEmpty()) {
+        markGlobalTxEnd(event);
+      }
+    }
+  }
+
+  private String getUniqueEventId(TxEvent event) {
+    return event.globalTxId() + "_" + event.localTxId();
```

Comment:
the map is based on globalTxId. why use globalTxId again?

Reply:
That's a good point. Use localTxId is enough.